sql/src/main/java/io/crate/expression/scalar/arithmetic/TruncFunction.java

```diff
@@ -0,0 +1,112 @@
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar.arithmetic;
+
+import io.crate.data.Input;
+import io.crate.expression.scalar.ScalarFunctionModule;
+import io.crate.expression.symbol.FuncArg;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.FunctionImplementation;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.FunctionResolver;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.params.FuncParams;
+import io.crate.metadata.functions.params.Param;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.List;
+import java.util.function.Function;
+
+
+public final class TruncFunction {
+
+    public static final String NAME = "trunc";
+    private static final FuncParams SIGNATURE = FuncParams
+        .builder(Param.NUMERIC)
+        .withVarArgs(Param.INTEGER)
+        .limitVarArgOccurrences(1)
+        .build();
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(NAME, new FunctionResolver() {
+
+            @Override
+            public List<DataType> getSignature(List<? extends FuncArg> funcArgs) {
+                return SIGNATURE.match(funcArgs);
+            }
+
+            @Override
+            public FunctionImplementation getForTypes(List<DataType> types) throws IllegalArgumentException {
+                int argsLen = types.size();
+                assert argsLen > 0 && argsLen <= 2 : "number of args must be up to 2";
+                DataType argType = types.get(0);
+                DataType returnTypeSingleArg = DataTypes.getIntegralReturnType(argType);
+                FunctionInfo info;
+                if (1 == argsLen) {
+                    info = new FunctionInfo(
+                        new FunctionIdent(NAME, List.of(argType)),
+                        returnTypeSingleArg);
+                } else {
+                    info = new FunctionInfo(
+                        new FunctionIdent(NAME, List.of(argType, DataTypes.INTEGER)),
+                        DataTypes.DOUBLE);
+                }
+                return new Scalar<Number, Number>() {
```

Comment:
would it be simpler to replace it with `UnaryScalar` and `BinaryScalar`, so we can delegate null checks and input materialization to those classes.

Reply:
thought of that, but then 

```
public BinaryScalar(BinaryOperator<T> func, String name, DataType<T> type, Set<FunctionInfo.Feature> feature) {
        this.func = func;
        this.info = new FunctionInfo(new FunctionIdent(name, Arrays.asList(type, type)), type, FunctionInfo.Type.SCALAR, feature);
        this.type = type;
    }
```

I need to fine tune the info.

Reply:
we can add a new ctor like: 
```
BinaryScalar(BinaryOperator<T> func, String name, List<DataType<T>> argTypes, DataType<T> returnType)
```

but maybe too much for this PR

Reply:
I think, I'd prefer to use a dedicated `UnaryScalar` for the 1 argument case, with 2 args we could always use the `BigDecimal.valueOf(val).setScale(numDecimals, mode).doubleValue()` logic, even with precision `0` it should work as defined.
I'd use the concrete scalar impl only for the 2 args case. Adjusting the `BinaryScalar` to support different argtypes and return types violates a bit its semantic imho.